File: Assignment_7/collcount.fs
( - 
  - collcount.fs
  - Aleks McCormick
  - 2023/04/28
  - Spring 2023 CS 331
  - Assignment 7 Exercise 2
  -   The same thing as the collatzCounts function from
  -      Assignment 5, but this time writen in Forth
  - )

\ collcount
\ Takes a positive integer and returns the number of iterations
\       of the Collatz function required to take the input value
\       to 1
: collcount { n -- c }
    0
    begin
    \ loop based collatz function
    n 1 > while
        n 2 mod 0 = if
            n 2 / to n
        else
            3 n * 1 + to n 
        endif
        \ Increment counter variable c
        1 +
    repeat
;    

File: Assignment_7/collcount.pl
/*
 * collcount.pl
 * Aleks McCormick
 * 2023/04/29
 * Spring 2023 CS 331
 * Assignment 7 Exercise 4
 *   The same thing as Assignment 7 Exercise 2
 *      but written in Prolog
 */

% collcount\2
% collcount(+n, ?c)

%  Case when N is 1
collcount(1, 0). 

% Case when n is even
collcount(N, C) :-
    N > 1,              % Verify N > 1
    N mod 2 =:= 0,      % Verify N is even
    Nr is N / 2,        % Set Nr to N/2
    collcount(Nr, Cr),  % Recurse passing Nr and receiving Cr
    C is Cr + 1.        % Set C equal to Cr + 1 and return C

% Case when n is odd
collcount(N, C) :-
    N > 1,              % Verify N > 1
    N mod 2 =\= 0,      % Verify N is odd
    Nr is (3 * N + 1),  % Set Nr to 3N+1
    collcount(Nr, Cr),  % Recurse passing Nr and receiving Cr
    C is Cr + 1.        % Set C equal to Cr + 1 and return C
